worlds/metroidprime/Logic.py

```diff
@@ -0,0 +1,292 @@
+from BaseClasses import CollectionState
+from .data.RoomNames import RoomName
+from .Items import ProgressiveUpgrade, SuitUpgrade, get_progressive_upgrade_for_item
+from typing import TYPE_CHECKING, Optional
+
+if TYPE_CHECKING:
+    from . import MetroidPrimeWorld
+
+
+def has_required_artifact_count(
+    world: "MetroidPrimeWorld", state: CollectionState
+) -> bool:
+    required_count = world.options.required_artifacts.value
+    return state.has_group("Artifacts", world.player, required_count)
+
+
+def can_boost(world: "MetroidPrimeWorld", state: CollectionState) -> bool:
+    return state.has_all(
+        [SuitUpgrade.Morph_Ball.value, SuitUpgrade.Boost_Ball.value], world.player
+    )
+
+
+def can_bomb(world: "MetroidPrimeWorld", state: CollectionState) -> bool:
+    return state.has_all(
+        [SuitUpgrade.Morph_Ball.value, SuitUpgrade.Morph_Ball_Bomb.value], world.player
+    )
+
+
+def can_power_beam(world: "MetroidPrimeWorld", state: CollectionState) -> bool:
+    return state.has_any(
+        [SuitUpgrade.Power_Beam.value, ProgressiveUpgrade.Progressive_Power_Beam.value],
+        world.player,
+    )
+
+
+def can_power_bomb(world: "MetroidPrimeWorld", state: CollectionState) -> bool:
+    if world.options.main_power_bomb.value:
+        return state.has_all(
+            [SuitUpgrade.Morph_Ball.value, SuitUpgrade.Main_Power_Bomb.value],
+            world.player,
+        )

```

Comment:
I would prefer to see `world.options` entirely removed from all logic rules. Individual access rules can be called 10,000s of times over the course of generation. A world's options are effectively static from the start of generation, so checking `world.options` within access rules is a waste of CPU time.

I would recommend that whenever a `can_power_bomb` rule is to be set somewhere, that is when `world.options` is checked and then the appropriate rule depending on `if world.options.main_power_bomb.value` is set.

---

I recommend replacing lists in `state.has_all`/`has_any` with tuples or, better, creating the lists/tuples in advance and put them into the module's globals. CPython pre-allocates memory for small tuples so they are faster to create, but putting them into the module's globals would mean they would only need to be created once, rather than on each call, which is even faster.

If you really want to push for performance, then tuples containing only string literals are the fastest because Python can precompute them into a single constant expression, e.g. `state.has_all(("Morph Ball", "Power Bomb (Main)"), world.player)`. But there's not much difference in performance compared to getting a list/tuple from the module's globals and you lose some ease of maintainability when using string literals.

In order of fastest to slowest (the first 3 are roughly the same performance, but are about 1.5x to 2x as fast as the last 3 when the length of the list/tuple is 2):
| Description | Code |
| --- | --- |
| The tuple contains only literals, so Python can constant fold it into a single constant without needing to build a new tuple. | `state.has_all(("Morph Ball", "Power Bomb (Main)"), world.player)` |
| The tuple has been created once in advance and stored on the module, so it has to be loaded from the module globals. | At the module level:<br>`power_bomb_main_items = (SuitUpgrade.Morph_Ball.value, SuitUpgrade.Main_Power_Bomb.value)`<br>Inside the function:<br>`state.has_all(power_bomb_main_items), world.player)` |
| Same as the previous, but with a list. Exactly the same instructions inside the function, but I'm putting the list lower because lists are barely slower to iterate and use slightly more memory than tuples. In most cases, this difference compared to tuples can be ignored. | At the module level:<br>`power_bomb_main_items = [SuitUpgrade.Morph_Ball.value, SuitUpgrade.Main_Power_Bomb.value]`<br>Inside the function:<br>`state.has_all(power_bomb_main_items), world.player)` |
| Now things start to get slower. Everything below here also gets noticeably slower the longer the list/tuple is. | |
| The item names are stored on the module in advance, so each one has to be loaded from the module's globals, and then a new tuple is created | At the module level:<br>`morph_ball = SuitUpgrade.Morph_Ball.value`<br>`main_power_bomb = SuitUpgrade.Main_Power_Bomb.value`<br>Inside the function:<br>`state.has_all((morph_ball, main_power_bomb)), world.player)` |
| The items names are constant string literals, so two constants are loaded and then a new list is created.<br>NOTE: Loading constants is faster than loading globals, so as the list gets longer, this will outperform the previous option.<br>NOTE: In some cases, Python can determine that it can replace lists containing only constants with a tuple constant expression, e.g. `for x in ["first", "second"]):` can be replaced with `for x in ("first", "second"):`, but passing the list to another function is not a case where Python can do this. | `state.has_all(["Morph Ball", "Power Bomb (Main)"], world.player)` |
| The item names are stored on the module in advance, so each one has to be loaded from the module's globals , and then a new list is created | At the module level:<br>`morph_ball = SuitUpgrade.Morph_Ball.value`<br>`main_power_bomb = SuitUpgrade.Main_Power_Bomb.value`<br>Inside the function:<br>`state.has_all([morph_ball, main_power_bomb]), world.player)` |

---

The amount of attribute lookups necessary to get each item name feels a little silly here, especially since the lookups have to be done every time `can_power_bomb` and the other functions are called. To get `SuitUpgrade.Morph_Ball.value`, `SuitUpgrade` has to be looked up from the module's globals, then the `Morph_Ball` attribute has to be looked up on that object and finally the `value` attribute has to be looked up on that object.

If you don't want to put string literals in your access rules, I'd recommend putting `morph_ball = SuitUpgrade.Morph_Ball.value` into the module's globals in advance. In this case, the entire list can be put into the module's globals, so this is more for rules that check a single item rather than a list of items.